readme.md
# CDTool v1.2b21

## Overview

A Windows CD reading and burning program that I wrote around 2008.
It is able to read and burn sectors to a CD-R or CD-RW.

Notable features:
- It supports writing in RAW DAO 2448 (2048+96) mode, meaning we can supply all of the sectors to be burnt, right from the start of the lead-in, all the way up to the end of the lead-out.
With other burning modes, certain parts, e.g. lead-in, 1st pregap or postgap and lead-out are generated by the burner itself.  It generates these sectors using the TOC that you send before burning.
- With certain drives, e.g. Lite-On ones, it is able to skip (unburnt) areas during the burn

The source code is written in Borland Delphi 7.


history.txt
17/12/2008. Version 1.2 beta 21:
- Fixed reading LBA from and higher than 404850 in View Sectors and CD to Image.

24/06/2007. Version 1.2 beta 20:
- Added reading of C2 error pointers.
- Fixed writing bug for SPT interface - this had prevented many writers from working, e.g. Pioneer.
- Fixed writing to ignore sense error "UNIT NOT READY, LONG WRITE IN PROGRESS" (code: 02/04/08).
- Added experimental writing raw DAO packed mode writing.

v1.2 beta 19 (09 April 2007)
- Fixed Error Recovery setting (all mode select commands were affected and are fixed).
- Changed Sector Viewer "Deinterleave subch data" to "Group subch bits".
- Display session nr. in TOC Viewer for full TOC mode.

v1.2 beta 18 (26 Oct 2006)
- Removed C2 offset correction when reading with subcode. Seems Premium didn't have a C2 offset to start with.

v1.2 beta 17
- Fixed ISRC reading bug - always reported not encoded on CD.
- Fixed reading "Read Disc" in ASPI mode.
- Fixed some bugs in error reporting.

31/03/2005. Version 1.2 beta 11:
- Beta 10 introduced a bug in the sending of main buffer data to drive, i.e. write and set error recovery commands did not work

29/03/2005. Version 1.2 beta 10a:
- Additions/changes
- Communication interface selection added

28/03/2005. Version 1.2 beta 10:
- Fixed: In TOC viewer, with read cd mode on and changing from different TOC reading modes did not clear previous data
- Fixed: (work around) windows 9x truncated form display (I believe it's a Delphi bug)
- Added: 'read track info'
- Added: 'read disc info'
- Added: Win32 DeviceIOControl with SPT interface for NT4/2K/XP/2K3

15/01/2004. Version 1.2 beta 9:
- Fixed: Sector viewer display was shifted by 1 one.
- Fixed: MSR option hanged - has now been fixed.  Also noticed that a spin up is required for measuring the beginning part of the CD - will try to implement this in next version.
- Fixed: Buffer underrun still occurred - even though a secondary cache was implemented.  The reason/fix: Image to CD sent only 1 sector per MMC command.  Now it sends more than 1 at a time, which prevents buffer underrun on faster writing speeds.  Note the skipping of sectors will still have this problem - since it sends only 1 sector per MMC command (work around: write at 1-4x only).
- Added: the option to use normal TOC (code 0000b) command to generate TOC file.  Some drives did not support the 'read Full TOC' command - which gives more complete data.
- Added: In View Sectors, added column width and different display formats for viewing the sector data
- Added: Added XP style.
- Added: In View TOC, unchecking the Read MCD, Read ISRC and Read CD mode will not clear the data read.
- Added: In View TOC, the ISRC are displayed as ASCII codes instead of as ASCII characters.
- Added: In Image to CD, test write option is added - note this doesn't work for some Plextor drives in RAW DAO mode.

01/10/2003. Version 1.2 beta 8:
- Fixed: a bug not all drives were listed.
- Fixed: recognising CD mode format of tracks.
- Fixed: generating single 2448 byte image file.
- Fixed: In 'CD To Image' the lead-in size produced a file that was smaller than what you entered, i.e. less 150 sectors.  This was okay before, because I gave wrong instructions on how to calculate lead-in size!  This has been fixed now.
- Added: a cyclic secondary buffer - this should now allow you to write at higher than 8x speed without buffer underrun.  At the moment it uses a default of 1/7 of total physical system RAM.
- Added: 2 extra features in to sector viewer: De-interlace sub-ch and yellow book scrambling options.
- Added: icon made by Sam donated to me from cdfreaks forum - thanks.
- Added: calculation of set pregap 1 to be written at LBA 0 in 'Image To CD'.
- Added: 'Read ATIP' in 'CD To Image' which calculates the lead-in length for a blank CDR/W.
- Added: CD mode 1 for lead-in generation into 'CD To Image'.
- Added: In 'ATIP' the buffer is now zeroed so that pressed CDs will show zeroed info, which makes more sense.  Some drives actually reported back with no error on pressed CDs when issuing the ATIP command.
- Added: Changed the way the program uses the MMC select10/sense10 commands - maybe more drives are supported now - haven't tested on stubborn drives.
- Added: Changed .LI1 and .LI2 extensions to more appropriate ones .LI and .PG1 respectively.
- Added: The generated .LO file now includes 2s of P sub-ch set to 0 at the start - in accordance with yellow book.
- Added: Added a newer way to skip sector positions.

28/05/2003. Version 1.2 beta 5:
- Added: some minor extra features - not enough to make a new version.  Added ability to specify starting position (expressed in LBA) when writing data to CD.  In 'Image To CD' the lead-in type for TOC is changed to more appropriate name and also added an extra item.

14/05/2003. Version 1.2 beta 4:
- Fixed: a bug from previous version:  Unable to write image to CD.

15/03/2003. Version 1.2 beta 3:
- Added: experimental generation of full lead-in TOC, lead-in pregap (first pregap of CD) and lead-out files - all of which also has generated headers and CRC/EDC/ECC if format is CDROM, lead-in sub-channels will be generated from TOC info read from CD and have CRC calculated
- Fixed: I think I've fixed the past the writing of 90:00:00 bug - still need more testing
- Fixed: SSP skipping sector bug.  The skip length for lead-in should now be positive values now
- Changed: .LED/.FLD extensions to .TU/TR
- Todo: Still no support for multi-session CDs - file generation will not work properly with this format - so don't use I didn't do much testing on this version so the above may not be correct.  Also I rushed a bit to bring this version out (still, It took me 2 weeks of work!) unused variables and copying and pasting of code were left in so file size is a little bit larger than usual.

16/12/2002. Version 1.2 beta 2:
- Fixed: bug of not using SSP file
- Added: measuring of sector reads/seeks (unit is recorded in micro seconds)

24/11/2002. Version 1.2:

SPC_inc.pas
{ Name:         SPC_DEFS.PAS
  File type:    Borland Delphi 4 include file.
  Description:  Misc SCSI definitions written for Borland Delphi 4.
                This source file is based on information given from
                the Adaptec's ASPI SDK and working draft version of
                SCSI-3 specification documents:
                     - SPC (SCSI Primary Commands)
                     - SAM (SCSI Architecture Mechanism)
  Notes:
  Date started: 19th Dec 1999.
  Developer:    Truong Hy.
}
{** SCSI peripheral device types (defined in SAM) **}
Const DTYPE_DIR_ACC =$00; //Direct access device, e.g. magnetic disk
Const DTYPE_SEQ_ACC =$01; //Sequential access device, e.g. magnetic tape
Const DTYPE_PRINTER =$02; //Printer device
Const DTYPE_PROC    =$03; //Processor device
Const DTYPE_WORM    =$04; //Write-once device, e.g. some optical disks
Const DTYPE_CDROM   =$05; //CDROM device
Const DTYPE_SCANNER =$06; //Scanner device
Const DTYPE_OPTI_MEM=$07; //Optical memory device, e.g. some optical disks
Const DTYPE_JUKEBOX =$08; //Medium changer device, e.g. juke box
Const DTYPE_COMM    =$09; //Communication device
Const DTYPE_RESL    =$0A; //Reserved low
Const DTYPE_RESH    =$1E; //Reserved high
Const DTYPE_UNKNOWN =$1F; //Unknown or no device type

{** Target device status values (defined in SAM) **}
Const DEV_ST_GOOD          =$00; //Target device did not detect an error
Const DEV_ST_CHKCOND       =$02; //Check condition status, error detected by device
                                 //check the returned sense data
Const DEV_ST_CONDMET       =$04; //Condition met (satisfied) for unlinked commands
Const DEV_ST_BUSY          =$08; //Target device is currently busy, cannot accept
                                 //any commands yet
Const DEV_ST_INTERM        =$10; //Linked command completed successfully
Const DEV_ST_INTERM_CONDMET=$14; //Condition met (satisfied) for linked commands
Const DEV_ST_RESCONF       =$18; //Device was reserved
Const DEV_ST_CMDTERM       =$22; //Command terminated
Const DEV_ST_QFULL         =$28; //Queue is full

{** SCSI MISC values **}
Const MAX_LUN_ID        =7;
Const DEFAULT_MAX_DEV_ID=7;
Const MAX_SCSI_LUN      =64;
Const MAX_SCSI_HA       =8;

{** SCSI-3 primary command codes **}
Const SCSI_CMD_TEST_UNIT_READY  =$00;
Const SCSI_CMD_REQ_SENSE        =$03;
Const SCSI_CMD_DEV_INQUIRY_CDB6 =$12;
Const SCSI_CMD_MODE_SELECT_CDB6 =$15;
Const SCSI_CMD_MODE_SENSE_CDB6  =$1A;
Const SCSI_CMD_MODE_SELECT_CDB10=$55;
Const SCSI_CMD_MODE_SENSE_CDB10 =$5A;

{** Request sense flags **}
{** Request sense response code **}
Const SEN_RESP_CURR_ERR   =$70; //Current errors
Const SEN_RESP_DEFER_ERR  =$71; //Deferred errors
Const SEN_RESP_VENDOR_SPEC=$7F; //Vendor specific
{** Request sense bit definitions **}
Const SEN_VALID  =$80;
Const SEN_FILEMRK=$80;
Const SEN_EOM    =$40;
Const SEN_ILI    =$20;
{** Request sense sense key definitions **}
Const SEN_KEY_NO_SEN     =$00; //No sense key info.
Const SEN_KEY_RECV_ERR   =$01; //Recovered or recoverable error/s.
Const SEN_KEY_NOT_READY  =$02; //Device not ready error.
Const SEN_KEY_MEDIUM_ERR =$03; //Medium error.
Const SEN_KEY_HARD_ERR   =$04; //Hardware error.
Const SEN_KEY_ILLEGAL_REQ=$05; //Illegal request, error/s in parameters or cmd.
Const SEN_KEY_UNIT_ATT   =$06; //Unit attention error.
Const SEN_KEY_DATA_PROT  =$07; //Data is protected for read or write access.
Const SEN_KEY_BLANK_CHK  =$08; //Blank or formatted medium encountered.
Const SEN_KEY_VEND_SPEC  =$09; //Vendor specific sense info.
Const SEN_KEY_COPY_ABORT =$0A; //Copy, compare or copy & verify aborted due to errors.
Const SEN_KEY_ABORTED_CMD=$0B; //Command aborted by device server.
Const SEN_KEY_EQUAL      =$0C; //Obsolete in SCSI-3 spec
Const SEN_KEY_VOL_OVERFLW=$0D; //Medium write capacity reached error.
Const SEN_KEY_MISCOMP    =$0E; //Source does not match data read error.
Const SEN_KEY_RESERVED   =$0F; //Reserved.
{** Request sense ASC/ASCQ flags **}

{** Mode select flags (applies to CDB6 and 10 versions) **}
Const MODE_SELECT_SP_NOSAVE  =$00;
Const MODE_SELECT_SP_SAVE    =$01;
Const MODE_SELECT_PF_VENDOR  =$00;
Const MODE_SELECT_PF_STANDARD=$10;

(** Mode sense flags **)
Const MODE_SENSE_DBD_BLK_DESC   =$00;
Const MODE_SENSE_DBD_NO_BLK_DESC=$08;
{** Page control flags. These should
be ORed with mode page code flags **}
Const MODE_SENSE_PC_CURRENT_VALS=$00;
Const MODE_SENSE_PC_CHGABLE_VALS=$40;
Const MODE_SENSE_PC_DEFAULT_VALS=$80;
Const MODE_SENSE_PC_SAVED_VALS  =$C0;
{** Mode page code flags. Rest of mode
page code flags are defined in device
specific class **}
Const MODE_SENSE_PGC_ALL   =$3F;

{** Misc types **}
Type
    T_array16=Array[1..16] Of Byte;
    T_array8=Array[1..8] Of Byte;
    T_array4=Array[1..4] Of Byte;


{** CDB for test unit ready command **}
Type T_test_unit_ready=Record
                             Cmd       : Byte;
                             Reserved1 : Byte;
                             Reserved2 : Byte;
                             Reserved3 : Byte;
                             Reserved4 : Byte;
                             Control   : Byte;
                       End;

{** CDB for request sense command **}
Type T_req_sense=Record
                       Cmd       : Byte;
                       Reserved1 : Byte;
                       Reserved2 : Byte;
                       Reserved3 : Byte;
                       Alloc_len : Byte;
                       Control   : Byte;
                 End;

{** Request sense return data format **}
Type T_sense_data=Record
                        ResponseCode   : Byte;
                        SegmentNo      : Byte;
                        Flags_SenseKey : Byte;
                        Info0          : Byte;
                        Info1          : Byte;
                        Info2          : Byte;
                        Info3          : Byte;
                        AddLen         : Byte;
                        ComSpecInfo0   : Byte;
                        ComSpecInfo1   : Byte;
                        ComSpecInfo2   : Byte;
                        ComSpecInfo3   : Byte;
                        ASC            : Byte;
                        ASCQ           : Byte;
                        FieldRepUCode  : Byte;
                        SenKeySpec15   : Byte;
                        SenKeySpec16   : Byte;
                        SenKeySpec17   : Byte;                        
                  End;

{** CDB for device inquiry CDB6 command **}
Type T_dev_inq_CDB6=Record
                          Cmd           : Byte;
                          CmdDt_EVPD    : Byte;
                          Pg_or_Op_code : Byte;
                          Reserved1     : Byte;
                          Len           : Byte;
                          Control       : Byte;
                    End;

{** Device inquiry standard return data format **}
Type T_dev_inq_std_data=Record
                              Dev_Qual_Type : Byte;
                              RMB           : Byte;
                              ConformVer    : Byte;
                              SupportFlags  : Byte;
                              AddLen        : Byte;
                              Reserved1     : Byte;
                              Flags1        : Byte;
                              Flags2        : Byte;
                              VendorID      : T_array8;
                              ProdID        : T_array16;
                              ProdRev       : T_array4;
                        End;

{** CDB for mode select CDB6 command **}
Type T_mode_select_CDB6=Record
                              Cmd            : Byte;
                              PF_SP          : Byte;
                              Reserved1      : Byte;
                              Reserved2      : Byte;
                              Param_list_len : Byte;
                              Control        : Byte;
                        End;

{** CDB for mode select CDB10 command **}
Type T_mode_select_CDB10=Record
                               Cmd                : Byte;
                               PF_SP              : Byte;
                               Reserved1          : Byte;
                               Reserved2          : Byte;
                               Reserved3          : Byte;
                               Reserved4          : Byte;
                               Reserved5          : Byte;
                               Param_list_len_MSB : Byte;
                               Param_list_len_LSB : Byte;
                               Control            : Byte;
                         End;

{** CDB for mode sense CDB6 command **}
Type T_mode_sense_CDB6=Record
                             Cmd            : Byte;
                             DBD            : Byte;
                             PC_PageCode    : Byte;
                             Reserved1      : Byte;
                             Param_list_len : Byte;
                             Control        : Byte;
                       End;

{** CDB for mode sense CDB10 command **}
Type T_mode_sense_CDB10=Record
                              Cmd                : Byte;
                              DBD                : Byte;
                              PC_PageCode        : Byte;
                              Reserved1          : Byte;
                              Reserved2          : Byte;
                              Reserved3          : Byte;
                              Reserved4          : Byte;
                              Param_list_len_MSB : Byte;
                              Param_list_len_LSB : Byte;
                              Control            : Byte;
                        End;

{** Generic mode parameter header for mode select/sense CDB6 **}
Type T_mode_param_hdr_6=Record
                              Mode_data_len      : Byte;
                              Medium_type        : Byte;
                              Dev_specific_param : Byte;
                              Block_desc_len     : Byte;
                        End;

{** Generic mode parameter header for mode select/sense CDB10 **}
Type T_mode_param_hdr_10=Record
                               Mode_data_len_MSB  : Byte;
                               Mode_data_len_LSB  : Byte;
                               Medium_type        : Byte;
                               Dev_specific_param : Byte;
                               Reserved1          : Byte;
                               Reserved2          : Byte;
                               Block_desc_len_MSB : Byte;
                               Block_desc_len_LSB : Byte;
                         End;

{** Generic (except direct access devices) mode parameter block descriptor **}
Type T_mode_param_block_desc=Record
                                   Density_code      : Byte;
                                   N_blocks_MSB      : Byte;
                                   N_blocks_MidByte  : Byte;
                                   N_blocks_LSB      : Byte;
                                   Reserved1         : Byte;
                                   Block_len_MSB     : Byte;
                                   Block_len_MidByte : Byte;
                                   Block_len_LSB     : Byte;
                             End;

{** Generic mode parameter header+block descriptor for mode select/sense CDB6 **}
Type T_mode_param_hdr_block_desc_6=Record
                                         Hdr        : T_mode_param_hdr_6;
                                         Block_desc : T_mode_param_block_desc;
                                   End;

{** Generic mode parameter header+block descriptor for mode select/sense CDB10 **}
Type T_mode_param_hdr_block_desc_10=Record
                                         Hdr        : T_mode_param_hdr_10;
                                         Block_desc : T_mode_param_block_desc;
                                   End;
